cohortextractor/tpp_backend.py

```diff
@@ -2720,6 +2721,92 @@ def patients_outpatient_appointment_date(
               OPA.Patient_ID
             """
 
+    def patients_with_value_from_file(self, f_path, returning=None, type_=None):
+        if not is_csv_filename(f_path):
+            raise TypeError(f"Unexpected file type {f_path}")
+
+        if returning is not None and type_ is None:
+            # StudyDefinition.get_pandas_csv_args() should have raised an error before
+            # we reach here.
+            raise TypeError("If `returning` is passed, then `type_` must be passed")
+
+        if returning is None and type_ is not None:
+            raise TypeError("If `type_` is passed, then `returning` must be passed")
+
+        if returning is None and type_ is None:
+            # If we set these here, then we don't need to keep track of which is None
+            # later.
+            returning = "value"
+            type_ = "bool"
+
+        # Fail before reading the CSV file
+        if type_ == "bool":
+            column_type = "INT"
+        elif type_ == "date":
+            column_type = "DATE"
+        elif type_ == "str":
+            column_type = "VARCHAR(MAX)"
+        elif type_ == "int":
+            column_type = "INT"
+        elif type_ == "float":
+            column_type = "FLOAT"
+        else:
+            # StudyDefinition.get_pandas_csv_args() should have raised an error before
+            # we reach here.
+            raise TypeError(f"Unexpected type {type_}")
+
+        # Which columns of the CSV file should we read?
+        usecols = ["patient_id"]
+        if returning != "value":
```

Comment:
That's certainly not inconceivable 🙂. To recap: If a user called `patients.with_value_from_file("records.csv", returning="value", type_="int")`, then their `value` column would be ignored. For this reason, the string `"value"` is a poor choice for the name of the dummy column. A dunder string, such as `"__value__"`, would be a better choice.

If a dunder string were a better choice, then it would appear in the final `SELECT` statement as the value of the `returning` variable:

https://github.com/opensafely-core/cohort-extractor/blob/c7021dcc78eafa82de5763b23cae5e4ca5bb380c/cohortextractor/tpp_backend.py#L2795-L2803

This seems undesirable. Would you suggest removing the dunder? Elsewhere, we use a `value` column to contain a boolean flag; `patients_all`, for example.

Reply:
> This seems undesirable

I don't really see any problem with that, assuming it's a valid SQL column name. Alternatively, something like `boolean_flag` would probably be OK too. It's just that `value` is so generic and "obvious" that I think there's a reasonable chance of something using it as a real column name.

Reply:
The column has to be named `value`, doesn't it? This code path is executed by `patients.which_exist_in_file`, which is used to create a study definition's population from the `patient_id` column in the underlying CSV file (see, for example, `test_patients_which_exist_in_file`). In this scenario, `TPPBackend.get_queries` gets an instance of `ColumnExpression` that wraps `"population.value"`; the column name `value` is the default (268).

https://github.com/opensafely-core/cohort-extractor/blob/3e59e90887031a9266005f7afb6fb7acb202d2ae/cohortextractor/tpp_backend.py#L265-L302

Alternatively, this column could be called `__value__` if we added an `AS` clause and removed the dunders.

Reply:
I'm unsure whether to continue to use `value` as the column name, with the limitation that the underlying CSV file shouldn't contain a column named `value`, or to refactor.

Reply:
If we rearrange the logic, then we can handle the case where a user wishes to read a column called `"value"` and also the case where a user wishes to call `patients.which_exist_in_file` 🙂 (i.e. where we create a new column called `value`, that is used by `TPPBackend.get_queries`). See 93bd638.